File: packages/types/src/live.ts
import { z } from "zod";
import { browserSchema, deviceTypeSchema } from "./bookmark";

// ── Live Sessions ("open tabs") ─────────────────────────────────────────────
// An ephemeral, per-device checkpoint of the tabs a browser has open. Kept in a
// separate table from saved sessions (never exported, TTL-reaped), so a live
// window can never be mistaken for a saved one. See docs/features/live-sessions.md.

/** One open tab within a live window. */
export const liveTabSchema = z.object({
  // Permissive like sessionTabSchema — XSS is defended at the render layer. Unlike
  // sessions, these URLs are ALSO sanitized at capture time in the extension
  // (§5.3), because render-layer defense does nothing against exfiltration.
  url: z.string(),
  title: z.string().max(300).default(""),
  favIconUrl: z.string().nullish(),
  active: z.boolean().optional(),
  /** URL reduced to its origin because the path/query looked credential-bearing (§5.3). */
  redacted: z.boolean().optional(),
});
export type LiveTab = z.infer<typeof liveTabSchema>;

/** One open browser window. `windowId` is display grouping only — never a key (§4.2.2). */
export const liveWindowSchema = z.object({
  windowId: z.number().int(),
  focused: z.boolean().optional(),
  tabs: z.array(liveTabSchema).max(100),
  /**
   * User-set display name for this window, overlaid onto list/stream responses
   * from a server-side per-device store (keyed by windowId). Absent/null = fall
   * back to the default "Window N" label. Never pushed by devices — it's a
   * viewer-side override, so the capture path leaves it undefined.
   */
  name: z.string().nullish(),
});
export type LiveWindow = z.infer<typeof liveWindowSchema>;

/**
 * POST /api/live body — a checkpoint push from one device. `capturedAt` is
 * client-clock DISPLAY METADATA only: never freshness, never ordering (§4.4).
 * `windows` absent = heartbeat (leave the mirror alone); `windows: []` = every
 * window is closed (do write the empty array). The two are different (§4.3).
 */
export const pushLiveStateSchema = z.object({
  deviceId: z.string().uuid(),
  label: z.string().max(80).optional(),
  browser: browserSchema.default("other"),
  device: deviceTypeSchema.default("other"),
  os: z.string().max(40).nullish(),
  capturedAt: z.string().datetime(),
  hiddenTabCount: z.number().int().min(0).default(0),
  windows: z.array(liveWindowSchema).max(12).optional(),
});
export type PushLiveStateInput = z.infer<typeof pushLiveStateSchema>;

/** One device in the GET /api/live response. `lastSeenAgeSeconds` is server-computed (§4.4). */
export const liveDeviceSchema = z.object({
  deviceId: z.string(),
  label: z.string(),
  browser: browserSchema,
  device: deviceTypeSchema,
  os: z.string().nullable(),
  windows: z.array(liveWindowSchema),
  tabCount: z.number().int(),
  hiddenTabCount: z.number().int(),
  lastSeenAt: z.string().datetime(),
  lastSeenAgeSeconds: z.number().int(),
  /**
   * This device's "new windows join live sessions by default" policy. ABSENT =
   * true (the fail-safe default — an unconfigured/new device auto-shares new
   * windows). Only present-and-false means the user opted the device out, so its
   * windows created after that are not shared unless individually turned on in the
   * extension popup. Set from the web app (PATCH /live/:deviceId/settings) and
   * mirrored into the extension from the push response.
   */
  newWindowsShared: z.boolean().optional(),
});
export type LiveDevice = z.infer<typeof liveDeviceSchema>;

/** GET /api/live response. `enabled` lets a reader tell "off" from "no devices" (§4.5). */
export const listLiveResponseSchema = z.object({
  devices: z.array(liveDeviceSchema),
  enabled: z.boolean(),
  ttlHours: z.number().int(),
});
export type ListLiveResponse = z.infer<typeof listLiveResponseSchema>;

/** POST /api/live/settings body — the account-wide opt-in flag. Off purges (§5.7). */
export const updateLiveSettingsSchema = z.object({
  enabled: z.boolean(),
});
export type UpdateLiveSettingsInput = z.infer<typeof updateLiveSettingsSchema>;

/**
 * PATCH /live/:deviceId/settings body — the per-device "new windows join live
 * sessions by default" policy. Off means windows opened on that device after the
 * change are NOT shared unless individually turned on in the extension popup.
 */
export const updateLiveDeviceSettingsSchema = z.object({
  newWindowsShared: z.boolean(),
});
export type UpdateLiveDeviceSettingsInput = z.infer<typeof updateLiveDeviceSettingsSchema>;

/**
 * POST /live 200 response. `enabled` echoes the account flag; `newWindowsShared`
 * is this device's new-window policy (absent = the default "true"), returned on
 * every push so the extension mirrors the policy with no extra request.
 */
export const pushLiveResponseSchema = z.object({
  ok: z.literal(true),
  enabled: z.literal(true),
  newWindowsShared: z.boolean().optional(),
});
export type PushLiveResponse = z.infer<typeof pushLiveResponseSchema>;

/**
 * PATCH /live/:deviceId/windows/:windowId body — rename one live window. An empty
 * string (after trim) CLEARS the override, restoring the default "Window N" label.
 */
export const renameLiveWindowSchema = z.object({
  name: z.string().trim().max(80),
});
export type RenameLiveWindowInput = z.infer<typeof renameLiveWindowSchema>;

File: apps/macos/Sources/Models/Live.swift
import Foundation

/// Live Sessions — the ephemeral "what's open right now" mirror served by the
/// dedicated live server (a different origin from the API; see
/// `packages/types/src/live.ts` for the canonical schemas).

/// One open tab within a live window.
struct LiveTab: Codable, Hashable, Sendable {
    var url: String
    var title: String?
    var favIconUrl: String?
    var active: Bool?
    /// URL reduced to its origin because the path looked credential-bearing.
    var redacted: Bool?
}

/// One open browser window. `windowId` is display grouping only — never a key.
struct LiveWindow: Codable, Hashable, Sendable {
    var windowId: Int
    var focused: Bool?
    var tabs: [LiveTab]
    /// User-set display name; absent/null = the default "Window N" label.
    var name: String?
}

/// One device currently mirroring tabs. `lastSeenAgeSeconds` is server-computed
/// — never derive freshness from client clocks.
struct LiveDevice: Codable, Hashable, Sendable, Identifiable {
    var deviceId: String
    var label: String
    var browser: String
    var device: String
    var os: String?
    var windows: [LiveWindow]
    var tabCount: Int
    var hiddenTabCount: Int
    var lastSeenAt: String
    var lastSeenAgeSeconds: Int
    /// This device's "new windows join live sessions by default" policy.
    /// ABSENT = true (the fail-safe default); only present-and-false means the
    /// user opted the device out.
    var newWindowsShared: Bool?

    var id: String { deviceId }
}

/// `GET {liveBase}/live` — also the payload of every SSE `state` frame.
/// `enabled` distinguishes "sharing is off" from "on, but no devices".
struct ListLiveResponse: Codable, Sendable {
    var devices: [LiveDevice]
    var enabled: Bool
    var ttlHours: Int
}

extension LiveTab {
    var openableURL: URL? {
        guard let url = URL(string: url), let scheme = url.scheme?.lowercased(),
              scheme == "http" || scheme == "https"
        else { return nil }
        return url
    }

    var displayTitle: String {
        if let title, !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return title }
        return url
    }

    var faviconURL: URL? {
        if let favIconUrl, let url = URL(string: favIconUrl), url.scheme != nil { return url }
        guard let host = openableURL?.host() else { return nil }
        return URL(string: "https://\(host)/favicon.ico")
    }
}

extension LiveDevice {
    /// "Active" = pushed within the last ~75s (the extension checkpoints well
    /// inside that); anything older renders as stale-but-still-useful.
    var isActive: Bool { lastSeenAgeSeconds < 75 }

    /// Folded behind the "Inactive devices" disclosure (like web/mobile):
    /// nothing pushed for 15 minutes — the browser is probably closed.
    var isInactive: Bool { lastSeenAgeSeconds >= 900 }

    var freshnessText: String {
        if isActive { return "active now" }
        let seconds = lastSeenAgeSeconds
        if seconds < 3600 { return "\(seconds / 60)m ago" }
        if seconds < 86_400 { return "\(seconds / 3600)h ago" }
        return "\(seconds / 86_400)d ago"
    }

    var symbolName: String {
        switch device {
        case "laptop": "laptopcomputer"
        case "desktop": "desktopcomputer"
        case "mobile": "iphone"
        case "tablet": "ipad"
        default: "display"
        }
    }
}

extension LiveWindow {
    /// The user-set name, else the conventional positional label.
    func displayName(at index: Int) -> String {
        if let name, !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { return name }
        return "Window \(index + 1)"
    }
}
